docs: refactor project configuration docs (#423)
mineiros-io/terramate: docs/project-config.md

# Project Configuration

Terramate does not depend on user configuration and comes with a set of sensible defaults.

> Configurations are valid projec-wide and can be defined only once at the project root.

Reviewer: ```suggestion
Configurations are valid project-wide and can be defined only once at the project root.
```

All project configurations are defined within the `terramate` block.

## The `terramate.required_version` Attribute

Required version is defined by the attribute `terramate.required_version` attribute
where `required_version` accepts a version constraint string,
which specifies which versions of Terramate can be used inside a Terramate project.

> We recommend to pin the exact version to use and update the config, when updating Terramate.

Reviewer 2: ```suggestion
We recommend pinning the exact version to use and updating the config when updating Terramate.
```

```hcl
terramate {
  # allow any Terramate v0.1.x version starting at v0.1.8
  required_version = "~> 0.1.8"
}
```

```hcl
terramate {
  # allow any Terramate v0.1.x version starting at v0.1.8
  required_version = ">= 0.1.8, < 0.2.0"
}
```

> If no `terramate.required_version` is defined on a project, no versioningss
> check will be performed. If a version constraint is defined, running `terramate` with an
> incompatible version will result in a fatal error.

Reviewer 2: ```suggestion
No versioning check will be performed if no `terramate.required_version` is defined on a project. If a version constraint is defined, running `terramate` with an incompatible version will result in a fatal error.
```

List of configurations:

- `terramate.config.git.default_remote` - Set a default remote origin to use for git related actions. The default is `"origin"`.
- `terramate.config.git.default_branch` - Set a default branch to use for git related actions. The default is `"main"`.

### The `terramate.config.run` Block

Configuration for the `terramate run` command can be set in the `terramate.config.run` block.

#### The `terramate.config.run.env` Block

> In `terramate.config.run.env` block a map of environment variables can be defined that will be set when running command using `terramate run`.

Reviewer 2: ```suggestion
In `terramate.config.run.env` block a map of environment variables can be defined that will be set when running a command using `terramate run`.
```

}
```

> In addition Globals (`global.*`) and Metadata (`terramate.*`) are available and are evaluated lazy withing the stack context the commands are executed in.

Reviewer: In the computer theory books, the term is called *lazy evaluation* and *lazy evaluated* is used in past sentences.

The `env` namespace is meant to give access to the host environment variables,
it is read-only, and is only available when evaluating
`terramate.config.run.env` blocks.

> This means that any attributes defined

Reviewer 2: ```suggestion
Any attributes defined
```

on `terramate.config.run.env` blocks won't affect the `env` namespace.

You can have multiple `terramate.config.run.env` blocks defined on different

> files, but variable names can NOT be defined twice.

Reviewer 2: ```suggestion
files, but variable names can **not** be defined twice.
```
